jetty-security/src/main/java/org/eclipse/jetty/security/authentication/ConfigurableSpnegoAuthenticator.java

```diff
@@ -0,0 +1,232 @@
+    {
+        return _authenticationDuration;
+    }
+
+    /**
+     * <p>Sets the duration of the authentication.</p>
+     * <p>A negative duration means that the authentication is only valid for the current request.</p>
+     * <p>A zero duration means that the authentication is valid forever.</p>
+     * <p>A positive value means that the authentication is valid for the specified duration.</p>
+     *
+     * @param authenticationDuration the authentication duration
+     */
+    public void setAuthenticationDuration(Duration authenticationDuration)
+    {
+        _authenticationDuration = authenticationDuration;
+    }
+
+    @Override
+    public Authentication validateRequest(ServletRequest req, ServletResponse res, boolean mandatory) throws ServerAuthException
+    {
+        if (!mandatory)
+            return new DeferredAuthentication(this);
+
+        HttpServletRequest request = (HttpServletRequest)req;
+        HttpServletResponse response = (HttpServletResponse)res;
+
+        HttpSession httpSession = request.getSession(false);
+        if (httpSession != null)
+        {
+            UserIdentityHolder holder = (UserIdentityHolder)httpSession.getAttribute(UserIdentityHolder.ATTRIBUTE);
+            if (holder != null)
+            {
+                UserIdentity identity = holder._userIdentity;
+                if (identity != null)
+                {
+                    Duration authnDuration = getAuthenticationDuration();
+                    if (!authnDuration.isNegative())
+                    {
+                        boolean expired = !authnDuration.isZero() && Instant.now().isAfter(holder._validFrom.plus(authnDuration));
+                        // Allow non-GET requests even if they're expired, so that
+                        // the client does not need to send the request content again.
+                        if (!expired || !HttpMethod.GET.is(request.getMethod()))
+                            return new UserAuthentication(getAuthMethod(), identity);
+                    }
+                }
+            }
+        }
+
+        String header = request.getHeader(HttpHeader.AUTHORIZATION.asString());
+        String spnegoToken = getSpnegoToken(header);
+
+        // The client has responded to the challenge we sent previously.
+        if (header != null && spnegoToken != null)
+        {
+            SpnegoUserIdentity identity = (SpnegoUserIdentity)login(null, spnegoToken, request);
+            if (identity != null)
+            {
+                if (identity.isEstablished())
+                {
+                    if (!DeferredAuthentication.isDeferred(response))
+                    {
+                        if (LOG.isDebugEnabled())
+                            LOG.debug("Sending final challenge");
+                        SpnegoUserPrincipal principal = (SpnegoUserPrincipal)identity.getUserPrincipal();
+                        setSpnegoToken(response, principal.getEncodedToken());
```

Comment:
So this is just an optional step?
If a Spnego token can only be used once, what is the purpose of this token on the response?  Is it just to show that the response was authenticated?    A comment here to describe what the return token is and why it is OK not to send it if auth is deferred would be good.

Reply:
In case a client sent (preempted) the `Authorization` token, _and_ the request is for a non-protected resource (so that `validateRequest()` is not called), _and_ eventually the application calls some API that triggers authentication and `validateRequest()` is called, _and_ the identity is established, then I guess we should send the token in the response.

We should remove the check for deferred authentication, and always send the response token back.

The response token must be sent so that the client too can establish a GSS context (that was only initiated when the client sent the first token). 
Clients may decide to only forward to the application responses for which the GSS context has been fully established, so we need to send the token back in the response.